app/core/effect.ts

```diff
@@ -1,13 +1,68 @@
 import { PokemonEntity } from "./pokemon-entity"
+import { Item } from "../types/enum/Item"
+import { Effect as EffectEnum } from "../types/enum/Effect"
+
+type EffectOrigin = EffectEnum | Item
 
 export abstract class Effect {
+  origin?: EffectOrigin
   apply: (entity: PokemonEntity) => void
-  constructor(effect: (entity: PokemonEntity) => void) {
+  constructor(effect: (entity: PokemonEntity) => void, origin?: EffectOrigin) {
     this.apply = effect
+    this.origin = origin
   }
 }
 
 // item effect applied on fight start of after stealing/obtaining an item
 export class OnItemGainedEffect extends Effect {}
 
-export class OnItemRemovedEffect extends Effect {}
\ No newline at end of file
+export class OnItemRemovedEffect extends Effect {}
+
+export class PeriodicEffect extends Effect {
+  intervalMs: number
+  timer: number
+  count: number
+
+  constructor(
+    effect: (entity: PokemonEntity) => void,
+    intervalMs: number,
+    origin?: EffectOrigin
+  ) {
+    super(effect, origin)
+    this.intervalMs = intervalMs
+    this.timer = intervalMs
+    this.count = 0
```

Comment:
If the goal is to reduce unnecessary static property duplication, I think this should be moved to the effects that use it since not every periodic effect needs a count. It's incremented at the same time the constructed method is called, so it should be fine to use only on classes that need it.

Reply:
Maybe, i hesitated. I could see it useful to have on generic PeriodicEffect

Reply:
Why not extend periodic effects to another abstract class that contains countable periodic effects? That way you prevent code duplication and also prevent unnecessary properties being on some classes.

Reply:
felt like overkill